### PR TITLE
Metric Tag Count for Points/Histograms/Spans

### PR DESCRIPTION
**Pull request by joannak0**
New metric in the proxy to find out how many point/histogram/span tags user is emitting

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java [hunk 1]
```diff
@@ -67,6 +69,8 @@
     MetricsRegistry registry = setupMetrics ? Metrics.defaultRegistry() : LOCAL_REGISTRY;
     this.receivedPointLag = registry.newHistogram(new MetricName(handlerKey.toString() +
         ".received", "", "lag"), false);
+    this.pointTagCount = registry.newHistogram(new MetricName(handlerKey.toString(), "", "pointTagCount"), false);
```
Comment (sushantdewan123) [inline proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:72]: rename this to 
pointTagCount.received

File: proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java [hunk 2]
```diff
@@ -67,6 +69,8 @@
...
+    this.histogramTagCount = registry.newHistogram(new MetricName(handlerKey.toString(), "", "histogramTagCount"), false);
```
Comment (sushantdewan123) [inline proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:73]: rename this to 
histogramTagCount.received

File: proxy/src/main/java/com/wavefront/agent/handlers/SpanHandlerImpl.java [hunk 3]
```diff
@@ -52,6 +56,8 @@
     this.validationConfig = validationConfig;
     this.validItemsLogger = validItemsLogger;
     this.dropSpansDelayedMinutes = dropSpansDelayedMinutes;
+    this.spanTagCount = Metrics.defaultRegistry().newHistogram(new MetricName(handlerKey.toString(), "",
+        "spanTagCount"), false);
```
Comment (sushantdewan123) [inline proxy/src/main/java/com/wavefront/agent/handlers/SpanHandlerImpl.java:60]: rename this to 
spanTagCount.received

**Review by sushantdewan123 — CHANGES_REQUESTED** [review]

File: proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java (hunk 1 above)
Comment (srujann) [reply proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:72]: probable points.tagCount.received.

Comment (srujann) [reply proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:72]: feel `points.tagCount.received` is slightly better

File: proxy/src/main/java/com/wavefront/agent/handlers/SpanHandlerImpl.java [hunk 4]
```diff
@@ -52,6 +56,8 @@
     this.validationConfig = validationConfig;
     this.validItemsLogger = validItemsLogger;
     this.dropSpansDelayedMinutes = dropSpansDelayedMinutes;
+    this.spanTagCount = Metrics.defaultRegistry().newHistogram(new MetricName(handlerKey.toString(), "",
```
Comment (srujann) [inline proxy/src/main/java/com/wavefront/agent/handlers/SpanHandlerImpl.java:59]: use the same pattern as `ReportPointHandlerImpl` to determine the registry to use.

**Review by srujann — CHANGES_REQUESTED** [review]

File: proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java (hunk 1 above)
Comment (basilisk487) [reply proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:72]: I have to disagree with both @srujann and @sushantdewan123 :) 
We should try to keep our naming at least somewhat consistent - I suggest naming it `.received.tagCount`, similar to `.received.lag` right above it. The final metric names would be like `~proxy.points.2878.received.tagCount.p99` etc.

File: proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java [hunk 5]
```diff
@@ -84,6 +88,13 @@ void reportInternal(ReportPoint point) {
       Histogram histogram = (Histogram) point.getValue();
       point.setValue(recompressor.apply(histogram));
     }
+
+    if (point.getValue() instanceof Histogram) {
+      histogramTagCount.update(point.getAnnotations().size());
```
[the remark this answers is not in the document]
Comment (srujann) [reply proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:93]: sounds good.

File: proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java (hunk 1 above)
Comment (sushantdewan123) [reply proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:72]: @basilisk487 the only difference here is that this metric is recorded pre-validation.
Hence I put in the keyword `raw` in there :)

Comment (sushantdewan123) [reply proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:72]: Are you okay with :)
**~proxy.points.2878.received.raw.tagCount.p99**

Comment (basilisk487) [reply proxy/src/main/java/com/wavefront/agent/handlers/ReportPointHandlerImpl.java:72]: what else would go under `received.raw.` metric tree? i would imagine `.received.tagCount` would be pre-validation, and if we need to, we could add `.sent.tagCount` for post-validation?

**Comment by joannak0** [comment]
@basilisk487 , please merge if you approve :) Thanks!

**Review by basilisk487 — APPROVED** [review]

**Pull request merged**